Linux-bionic (Android libc w/o Android interop) initial implementation

directhex opened the pull request:
Implement support for "Linux Bionic", a RID which builds a Linux runtime against an Android libc but without Android OS functions like JNI, Dalvik, Instrumentation, etc. This configuration has been requested by a team working on an IoT scenario.

* The new RID is linux-bionic-arm64/linux-bionic-x64.
* The build reuses the existing Android build dockerfiles and Helix queues
* Since we don't have JNI, we don't have Android Crypto - instead, use an Android build of OpenSSL. As an errata, it is incumbent upon the end user to supply this in their OS environment (we download one ad-hoc for testing purposes, which was compiled by Google)
* Some tests cannot be completed due to SELinux policies affecting the `shell` user executing tests via `adb shell` in the world-writable `/data/local/tmp` directory. We expect those tests would pass if they pass on Android, in a real-world IoT scenario with a custom OS (rather than recycling Google's OS images intended for phones)

Closes: #66027

directhex committed:
Initial linux-bionic implementation.

Closes: #66027

directhex committed:
Add some CI

directhex committed:
Fix the build

Commit:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Update compat

directhex committed:
Merge branch 'joshield/linux-bionic' of github.com:directhex/runtime into joshield/linux-bionic

marek-safar commented on src/libraries/Microsoft.NETCore.Platforms/src/runtimeGroups.props:
```diff
@@ -13,6 +13,10 @@
       <Architectures>x64;x86;arm;armel;arm64;s390x</Architectures>
     </RuntimeGroup>
+    <RuntimeGroup Include="linux-bionic">
+      <Parent>android</Parent>
```
I think this should be the other way around.

directhex committed:
WIP to make this smaller & use `-os android` as a base

directhex committed:
Merge branch 'joshield/linux-bionic' of github.com:directhex/runtime into joshield/linux-bionic

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Leftover crud

directhex commented:
Urgh, horrible merge failure means I spent hours cleaning up unneeded changes, and they reappeared anyway

directhex committed:
More cleanup

directhex committed:
Refreshed runtime compat after inverting Android parenthood

directhex commented:
New regression: SSL doesn't work because an Android System.Net.Security is being used instead of a Linux one, which causes it to forcibly load Android SSL instead of OpenSSL.

steveisok commented:
> New regression: SSL doesn't work because an Android System.Net.Security is being used instead of a Linux one, which causes it to forcibly load Android SSL instead of OpenSSL.

I don't think System.Net.Security is going to work on Android anyway. At least not in .NET 6.

directhex commented:
> > New regression: SSL doesn't work because an Android System.Net.Security is being used instead of a Linux one, which causes it to forcibly load Android SSL instead of OpenSSL.
> 
> I don't think System.Net.Security is going to work on Android anyway. At least not in .NET 6.

That's kinda the point. It worked yesterday when the build system assumed Bionic was Linux, and it's broken today now the build system thinks Bionic is Android, and I'm trying to figure out how to massage that in the build system to behave better (msbuild log viewer shows none of my properties in System.Net.Security.csproj are getting resolved, something odd is going on)

steveisok commented:
Tweak this and you should be ok: https://github.com/directhex/runtime/blob/78111b553b2546356130a1d7ed33c944f3a94e46/src/libraries/System.Net.Security/src/System.Net.Security.csproj#L14

directhex commented:
> Tweak this and you should be ok: https://github.com/directhex/runtime/blob/78111b553b2546356130a1d7ed33c944f3a94e46/src/libraries/System.Net.Security/src/System.Net.Security.csproj#L14

You'd think! And yet

![Screenshot from 2022-03-07 15-11-40](https://user-images.githubusercontent.com/308992/157110530-99707f5e-5c50-4097-bf78-13603984662a.png)

Hence "and something odd is going on"

directhex commented:
Looks like the problem is with the structured log viewer, I see more via `/v:diag`

directhex committed:
Fix build to use -os Linux as a base, not -os Android

directhex committed:
Another stupid merge failure from days ago

directhex commented:
Okay, so, need help from someone who understands the libraries build a little better here (@ericstj?)

* The build doesn't really support OS subgroups (i.e. `linux-musl`, `linux-bionic`), it's a gross hack where the system thinks it's actually the OSGroup OS (Linux for musl), some magic autodetection code in a few shell scripts tells it otherwise and causes changes in build behaviour, and the RID gets overridden post-hoc via `/p:OutputRID`. I've removed this assumption in a few places and parsed `/p:RuntimeOS`, because it's not possible to autodetect the difference between Android and linux-bionic based on anything in the rootfs.
* We still need a TargetOS (`--os`). The `linux-musl` build defaults to `linux`; in our case there's a case to be made for both `linux` and `android`. And we kinda need to be both, in a few places - System.Net.Security and System.Security.Cryptography we need the Linux version NOT the Android version, whereas System.Console we want the Android version. Do we want to add a separate TFM for linux-bionic? Use RuntimeOS conditionals in the relevant csproj? What's the least painful way to do this?

jkotas commented:
> The build doesn't really support OS subgroups (i.e. linux-musl, linux-bionic),

Correct. All linux OS subgroups differences have been contained in native shim under src/native/libs. We have identical managed binaries for all linux OS subgroups. It would be nice to keep it that way.

jkotas commented:
> whereas System.Console we want the Android version.

Is it really the case? I would expect that we want regular System.Console for the vanilla linux-bionic version.

steveisok commented:
> > whereas System.Console we want the Android version.
> 
> Is it really the case? I would expect that we want regular System.Console for the vanilla linux-bionic version.

I thought it made more sense to plug into logcat since it's the standard android way. That said, I don't have too strong of an opinion.

directhex commented:
> > whereas System.Console we want the Android version.
> 
> Is it really the case? I would expect that we want regular System.Console for the vanilla linux-bionic version.

Okay, System.Net.NetworkInformation then. The limits are enforced at the kernel/libc level, not the presentation layer level. We definitely want the same System.Net.NetworkInformation as Android, not the same as desktop Linux.

jkotas commented:
I see that the `System.Net.NetworkInformation` has some things (like reading from /proc file system) short-circuited for Android. I looks like a small size optimization. It should not be a big deal to drop that size optimization for bionic and just use the regular Linux System.Net.NetworkInformation.dll binary. We may need to tweak the error handling in the regular System.Net.NetworkInformation.dll Linux binary so it throws something that makes sense when the /proc file system is not available.

am11 commented:
> We may need to tweak the error handling in the regular System.Net.NetworkInformation.dll Linux binary so it throws something that makes sense when the /proc file system is not available.

I think it was recently improved in https://github.com/dotnet/runtime/pull/63696. AFAIK, .NET 7 runtime and SDK (https://github.com/dotnet/sdk/pull/22927) should work in environments with missing or misbehaving procfs.

directhex committed:
Accepting any runtimeos in corehost build breaks non-portable build

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Include runtime in correlation payload build

am11 commented on src/libraries/externals.csproj:
```diff
@@ -9,7 +9,7 @@
     <BinPlaceRuntime>false</BinPlaceRuntime>
     <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
-    <UseLiveBuiltDotNetHost Condition="'$(TargetArchitecture)' == 's390x' or '$(TargetArchitecture)' == 'armv6'">true</UseLiveBuiltDotNetHost>
+    <UseLiveBuiltDotNetHost Condition="'$(TargetArchitecture)' == 's390x' or '$(TargetArchitecture)' == 'armv6' or '$(RuntimeOS)' == 'linux-bionic'">true</UseLiveBuiltDotNetHost>
```
We are trying to (get away from legacy `RuntimeOS` and `__DistroRid` and) converge things on two things: either it is "identifier to restore (most of the) packages - `PackageRID`" or "identifier to create shipping packages - `OutputRid`" (latter seldom differs from former in cases like non-portable builds). In this case, this condition maybe better:
```suggestion
    <UseLiveBuiltDotNetHost Condition="'$(TargetArchitecture)' == 's390x' or '$(TargetArchitecture)' == 'armv6' or $(OutputRid.StartsWith('linux-bionic'))">true</UseLiveBuiltDotNetHost>
```

directhex replied:
Where is that model used within the repo? I was following the existing use of `'$(RuntimeOS)' == 'linux-musl'`

am11 replied:
> Where is that model used within the repo?

See line 44 below as an example.

In the effort to simplify number of ways to alter host/target info from command-line, we consolidated and cleaned up these properties. RuntimeOS == X conditions used to be all over the place in .NET Core 3.1 and before, we have combined all that and cleaned it up from various places. We use PackageRID or OutputRid as much as possible. The reason is we want to compute the host/target info once, in one place, for the entire build. Also, RuntimeOS comes from "outside" which is not the first preference; we want to determine the host and target by ourselves and reduce the support of number of overlapping command-line options which change the build.

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Generate and use on-device test runner script for Bionic

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Latest changes (actually runs on Helix, w/ new enough xharness)

directhex committed:
Gross workaround for ICU failure
On Android 29, the OS ICU library is miscompiled and does not find its own ICU data file. It is safe to override the directory (via an env var) to a non-existent folder, so force it to the Android 29 value and it'll be cleanly ignored on Android 31.

directhex committed:
Add missing file

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Fix heisenbug

directhex committed:
OpenSSL

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Missing test exclusion

Commit:
Support for building against OpenSSL 3 headers

Commit:
Use some Android parts for Bionic CoreLib

CoreLib is intentionally per-platform, shouldn't violate rules about per-platform shared libs

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Run in CI

This'll probably fail pending XHarness updates

directhex committed:
Use OutputRID not RuntimeOS for Bionic switches, as requested

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Deal with Android build failure

directhex committed:
Workaround for DNNE restore error

I haven't worked out why TargetsLinuxBionic doesn't work here

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

jkotas committed:
Update aka.ms link with NativeAOT optimization guide (#68151)

directhex committed:
Fix the build

maxkoshevoi committed:
[Group 9] Enable nullable annotations for `Microsoft.Extensions.Hosting.WindowsServices` (#67512)

* Annotate

* Update Win32.cs

directhex committed:
Get rid of debug spew

BruceForstall committed:
Enable multiple nested "no GC" region requests (#68105)

When doing fast tail call on arm32, we disable GC reporting.
But some IR nodes, like unrolled STORE_BLK, also disable GC
reporting. If one of those is within a fast tail call argument
setup region, we'll attempt to disable GC reporting twice. Since
we didn't keep track of nesting, we end up marking some of the
tail call region after the STORE_BLK as interruptible, leading
to be GC info in the argument area.

Change the enable/disable GC calls to keep a nesting count,
and only re-enable GC reporting when the count reaches zero.

Fixes #67046

jakobbotsch committed:
JIT ARM32: Force tailcall nullchecks with LR (#68107)

For interface tail calls we need to allocate registers for both the
target and a null-check. The register that can be used for the target is
highly constrained: it essentially only go into r12 as the arg registers
may be busy and that is the only volatile register left that will not be
overridden by the epilog. This leaves
just `LR` left for the null check, and LSRA does not seem to handle the
low amount of freedom too well (these are all allocated at the same
location, which might have something to do with it).

To help LSRA just force the null-check to happen with LR for fast
tailcalls, which is going to be killed going into the epilog anyway.

Fix #66563

tannergooding committed:
Exposing generic math support for unsigned right shift (#68096)

* Updating Microsoft.Net.Compilers.Toolset to v4.3.0-1.22215.4 for unsigned-right-shift

* Exposing operator >>> (Unsigned Right Shift)

* Ensure that SByte and Int16 have unsigned-right-shift operate on themselves

* Adding tests covering Unsigned Right Shift

maxkoshevoi committed:
[Group 9] Enable nullable annotations for `Microsoft.Extensions.Hosting.Systemd` (#67513)

Fix #43605

* Annotate

* Fix API compat check by removing MemberNotNullWhen attribute.

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

EgorBo committed:
SPMI: fix AV in getStringLiteral (#68161)

AaronRobinsonMSFT committed:
Support ByRefLike types as Generic parameters (#67783)

* Add flag for ByRefLike generic constraint

* Update ILAsm/ILDasm to support byreflike keyword

* Runtime and Libraries tests

* Cast defined int constant until managed API is approved.

dakersnar committed:
Update VSCode library debugging documentation (#67925)

tannergooding committed:
Expose APIs for getting the exponent and significand for IFloatingPoint types (#68094)

* Updating Half/Single/Double to consistently track the internal bitwise manipulation constants

* Expse APIs for getting the exponent and significand for IFloatingPoint types

* Expose tests for getting the exponent and significand for IFloatingPoint types

* Apply suggestions from code review

Co-authored-by: Drew Kersnar <dakersnar@me.com>

* Assert that decimal.Exponent is between 95 and 67, inclusive

Co-authored-by: Drew Kersnar <dakersnar@me.com>

stephentoub committed:
Fix compiled / source generated lazy loop stack handling (#68138)

When a lazy loop has completed iterations but then ends up failing, and the thing it's looping around has no backtracking constructs, we're incorrectly leaving some state associated with the loop on the backtracking stack.

cshung committed:
Lift the restriction that we can have at most 4k regions (#68168)

kant2002 committed:
Fix nullability warnings for NativeAOT CoreLib (#65137)

* Fix nullability warnings for NatveAOT CoreLib
Enable checks for CS8600
Contributes to dotnet/runtimelab#106

* Bunch of CS8602

* Apply PR feedback

* Fix compile error

* Fix errors

* Use notnull constraint

Commit:
Update dump template (#68185)

trylek committed:
Improve performance of test partitioning (#68172)

I have noticed a perf bug I made as part of the refactoring for
merged test wrapper support. When there are no merged wrappers
available due to test filtering (using the test / dir / tree
commands-line argument), %(MergedAssemblyParentFolders.Identity)
expands to an empty string at line 397 and AllRunnablePaths
effectively search the entire HDD for tests to remove. In my
case it reduced the running time of GetListOfTestCmds from 23
minutes to 3 milliseconds.

Thanks

Tomas

tannergooding committed:
Updating System.Runtime.InteropServices.NFloat to support generic math (#68141)

* Updating NFloat to implement IBinaryFloatingPointIeee754 and IMinMaxValue

* Adding tests for the NFloat generic math support

* Fixing NFloat to expose the APIs for getting the exponent and significand

* Fix the NFloat.GetSignificandByteCount implementation

pedrobsaila committed:
CheckValue is not a fast path for all callers (#68125)

* CheckValue is not a fast path for all callers

* Revert "CheckValue is not a fast path for all callers"

This reverts commit a24fef61ec5bf4ddc1be386781ed26021a656c26.

* fix the SetValue caller

jakobbotsch committed:
Fix very adventurous LIR::Use::GetDummyUse (#68131)

This only happened to work due to Named RVO and is UB in any case.

kant2002 committed:
Add links to docs (#68211)

That hopefully simplify path for brave person willing to add new target

LakshanF committed:
Fixing casing issue with TargetPath (#68184)

* Fixing casing issue with TargetPath

rolfbjarne committed:
[AOT] Don't set the 'CorrectedSynthesize' flag in the objc_imageinfo section. (#67441)

* [AOT] Don't set the 'CorrectedSynthesize' flag in the objc_imageinfo section.

According to Apple's headers, it's not used anymore, and ignored.

However, with Xcode 13.3, it seems Apple has re-used this value for something
else, and it's causing warnings in the linker:

> ld: warning: '[...]/arm64/Xamarin.iOS.dll.o' was built with class_ro_t pointer signing enabled, but previous .o files were not

Fixes https://github.com/mono/mono/issues/21450.

Ref: https://opensource.apple.com/source/objc4/objc4-818.2/runtime/objc-abi.h.auto.html

* Don't set the 'CorrectedSynthesize' flag when using LLVM either.

Bo98 committed:
Fix CLR_CMAKE_USE_SYSTEM_LIBUNWIND on macOS (#68116)

eiriktsarpalis committed:
Add JSON type hierarchies (#67961)

* Add JSON type hierarchies as a preview feature.

* fix incorrect preview features configuration in ref project

* Add check for conflicting custom type discriminator property names.

* Add null checks & tests to public APIs

* fix typo in constant name

* fix file ordering

* address feedback

* Remove preview features annotation from new APIs

Commit:
Fix issue #41084 - fix cmake CMP0068 warnings (#68095)

ViktorHofer committed:
Fix path in servicing doc (#68208)

jkotas committed:
Cleanup ProjectN references (#68201)

dakersnar committed:
Fix incorrect results of right shifting huge BigIntegers error reported in Issue #65633 (#67867)

* Fix #65633, multiplication overflow causing incorrect right shift results

* Add unit test reproducing incorrect right shift results for BigInteger (#65633)

* Fix inconsistent formatting

* Expand LargeNegativeBigIntegerShiftTest to validate values, expose private BigInteger members to tests

* Apply suggestions from code review

Co-authored-by: Tanner Gooding <tagoo@outlook.com>

* Restrict unit test to 64-bit processors

* Document reasoning for restricting the test to 64-bit

Co-authored-by: Dan Moseley <danmose@microsoft.com>

Co-authored-by: Tanner Gooding <tagoo@outlook.com>
Co-authored-by: Dan Moseley <danmose@microsoft.com>

vargaz committed:
[mono][jit] Add SIMD support for System.Numerics.Vector4. (#68098)

trylek committed:
Consistently use forward slashes in test exclusion lists (#68215)

jharjung committed:
Improving STJ source generator support for record types  (#68064)

* Improving STJ source generator support for record types

Improving the analyzer to not flag a warning about init-only
properties that are actually constructor parameters. In particular,
record types with positional parameters desugar to classes with both
a constructor and init-only properties (eg, to support "with"
expressions). Since these properties are deserialized via the
constructor and not the property setter, they support the fast path and
round trip just fine, so the diagonstic analyzer should not raise the
SYSLIB1037 warning.

Issue 58770

* Improving STJ source generator support for record types

Edit to fix missing semi colon in unit test's CreateCompiliation string

Issue 58770

* Case insensitive matching of props with ctor params

* Additional unit tests

* Test for source gen positional records roundtrip

Co-authored-by: Joe Harjung <jharjung@integratedlabor.com>

Commit:
Revert "Update dump template (#68185)" (#68224)

This reverts commit 205632116bd764abfe019564256908202082c9a9.

elinor-fung committed:
Expose ArrayMarshaller and PointerArrayMarshaller (#68173)

filipnavara committed:
Managed implementation of NTLM for Android and tvOS (#66879)

* Move MD4 implementation into Common/src/System/Net/Security

* Add minimal RC4 implementation

* WIP: Integrate managed Ntlm implementation into SocketsHttpHandler

Co-authored-by: Tomas Weinfurt <tweinfurt@yahoo.com>

* WIP: Makeshift tests for NTLM/Negotiate authentication

* Fix compilation, clean up some of the hashing

* Avoid using a temporary buffer

* Add computation of signing keys, sealing keys and mechListMIC

* Various cleanups

* Send SPN in target information

* Add some validation, mark spots with missing validation

* Clean up some of the memory offset manipulation

* Move NTLM version into static variable

* Add support for channel bindings, clean up

* Fix hash calculation in makeNtlm2Hash accidentally broken with last commit.
Read NegTokenResp explicitly.
Add mechListMIC reading and verification.

* Verify last authentication token in HTTP Negotiate authentication

* Address feedback

* Fix tvOS builds by making few methods static

* Enable System.Net.Security tests on Android and iOS

Co-authored-by: Tomas Weinfurt <tweinfurt@yahoo.com>
Co-authored-by: Steve Pfister <steve.pfister@microsoft.com>

AndyAyersMS committed:
JIT: fix issue with invalid operand size in OSR prolog on arm64 (#68198)

Need to use `emitActualTypeSize` for the operand size.

Fixes #68170.

olsaarik committed:
Two-phase matching algorithm for NonBacktracking (#68199)

* Switch to 2-phase matching in NonBacktracking

First phase now finds the true match end position.
The implicit .* is now a lazy .*? to prioritize the earliest match.
Third phase is now only run for subcaptures, which no longer needs to
find match end position.
Remove counter optimization that no longer applies with OrderedOr.
Fix a problem in SymbolicRegexInfo where begin/end anchors were
marked as line anchors.
Also remove dead fields from SymbolicRegexInfo.
Fix captures not being handled for empty matches at start of input.

* Improve comments for NonBacktracking

Especially fix comments for the new 2-phase match generation algorithm.

* Add a failing test for the earlier NonBacktracking

* Avoid transitions to deadends for capuring NFA

AndyAyersMS committed:
JIT: fix patchpoint offset encoding (#68202)

We can now have Tier0 locals at byte offsets, so rework how the offset
information is incoded in patchpoints to make this possible.

Closes #68194.

wfurt committed:
avoid allocating collection for intermediate certificates (#68188)

* avoid allocating collection for intermediate certificates

* fix NRE for custom trust

* feedback from review

kouvel committed:
Fix a race condition in the thread pool (#68171)

* Fix a race condition in the thread pool

There is a case where on a work-stealing queue, both `LocalPop()` and `TrySteal()` may fail when running concurrently, and lead to a case where there is a work item but no threads are released to process it. Fixed to always ensure that there's a thread request when there was a missed steal. Also when `LocalPop()` fails, the thread does not attempt to pop anymore and that can be an issue if that thread is the last thread to look for work items. Fixed to always check the local queue.

Fixes https://github.com/dotnet/runtime/issues/67545

stephentoub committed:
Add initial version of {Last}IndexOfAnyExcept (#67941)

* Add initial version of {Last}IndexOfAnyExcept

These are functional but not vectorized.  At least some of these should be vectorized for at least some data types subsequently, but that's a more intensive change.  Once that's in, we can update a few places to use these, e.g. Regex should end up using any of the overloads that are vectorized.

* Fix comments

jkotas committed:
Reflection.Emit cleanups (#67170)

* Delete AssemblyBuilder.CheckContext checks

This check is trying to ensure that the Types passed in are somewhat well-behaving by verifying that
Module and Assembly properties are non-null. This validation is not done consistently and it has very
value. There are many other ways custom types can misbehave.

Also, this saves allocations of the object[] arrays for the params argument.

* Simplify Exception.Source

Simplify Exception.Source to avoid Reflection.Emit dependencies in Exception implementation

* Delete unnecessary clone of CustomAttributeBuilder enumeration

* Factor out simple ModuleBuilder and TypeBuilder forwardering methods

* Delete ComVisible annotations in Mono

* Delete redundant FCall

jkotas committed:
Delete AssemblySpec::m_wszCodeBase (#68186)

* Delete AssemblySpec::m_wszCodeBase

This was only used for assembly loading using hosting APIs that can be short-circuited to load directly without going through AssemblySpec

* Delete unnecessary ownedFlags argument

BruceForstall committed:
Treat SuperPMI failures in AzDO pipelines as pipeline failures (#68196)

Now, any SuperPMI failure in the replay, asmdiffs, and checked/release
asmdiffs pipeplines is ignored, and doesn't cause AzDO to treat the entire
job as failed. The job does get a "!" icon, but that's only if you visit the
AzDO UI: the GitHub page on a PR doesn't see that failure.

Now that we have more experience with these pipelines, and we are using
replay and asmdiffs with JIT PRs, we want the failures to be more visible,
and treated as failures.

Note that asm diffs should not be treated as failures.

tannergooding committed:
Exposing MaxMagnitudeNumber, MaxNumber, MinMagnitudeNumber, and MinNumber (#68217)

* Exposing MaxMagnitudeNumber, MaxNumber, MinMagnitudeNumber, and MinNumber

* Ensure Max, MaxMagnitude, Min, and MinMagnitude tests are covering the relevant scenarios

* Moving the floating-point AssertEqual helper from the Math/MathF tests into AssertExtensions

* Adding tests covering MaxMagnitudeNumber, MaxNumber, MinMagnitudeNumber, and MinNumber

* Disabling a few Half tests due to an issue around -0

* Disabling a few Half tests due to an issue around +0

* Fixing 67993 as the test helper had a bug where it used the wrong target type

* Fixing up the doc comment for Min/MaxMagnitudeNumber and Min/MaxNumber to clarify it doesn't propagate NaN

huoyaoyuan committed:
Remove platform dependent compilation of System.Private.Uri (#68011)

* Use OperatingSystem.IsWindows() in Uri.

* Remove platform specific files.

* Remove platform specific compilation for System.Private.Uri.

akoeplinger committed:
[mono] Rename .md files that are not actually markdown (#68239)

See https://github.com/dotnet/runtime/pull/68221

ViktorHofer committed:
Exclude all markdown files from triggering CI (#68221)

* Exclude all markdown files from triggering CI

Since 08/09/2021 it's now possible to use wild cards in exclusion patterns: https://docs.microsoft.com/en-us/azure/devops/release-notes/2021/sprint-192-update#support-for-wild-cards-in-path-filters.

Fixing our exclude yml statements to ignore all markdown files in the repo and not just the ones under docs/.

carlossanlop committed:
Update 'main' branding to preview5 (#68236)

teo-tsirpanis committed:
Clean-up `System.Drawing.Common` and remove the Unix code. (#64623)

* Remove all Unix-specific files.

* Throw PNSE on non-Windows.

* Merge the Windows-specific files into their formerly cross-platform counterparts.

* Remove all mentions of Unix in the tests.

* Remove two always-on defines.

* Merge two item groups in the project file.
No reason to sort them; the list is already unsorted.
And rename a remaining formerly Windows-specific file.

* Remove the NoCOMWrappers files.

* Fail on unsupported platforms when a library is trying to be loaded.

* Fix compile errors.

* Small changes in the project file.

* Remove two meaningless asserts.

* Run BinaryFormatter tests on SDC types only on Windows.

* Use `[ThreadStatic]` in Gdip.ThreadData.

* Remove `TargetsAnyOS`.

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

* SDC project file clean-up

* Remove `!!` from System.Drawing.Common.

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

jakobbotsch committed:
Retype pointers to 'this' on stack as unmanaged pointers (#68074)

Remove one of the quirks added in the call args refactoring.

eiriktsarpalis committed:
Enable System.Text.Json tests on netfx (#63803) (#67007)

* Enable System.Text.Json tests on netfx (#63803)

* Enable System.Text.Json tests on netfx

* use NETFRAMEWORK define

* disable another two test cases after rebase

* disable last test case which repros only on CI for me

* add p2p only on netfx

* Remove hardcoded assembly qualified names

* Address feedback

Co-authored-by: Krzysztof Wicher <kwicher@microsoft.com>

stephentoub committed:
Overhaul when/where we check for Regex timeouts (#68146)

* Overhaul when/where we check for timeouts

Regex timeouts have never been about guaranteeing exact timeout handling; rather, they're about avoiding catastrophic backtracking.  As such, we already allow an O(n) amount of work in many cases between timeout checks.  This change formalizes that, such that we now check for a timeout at every place where we could do at least an O(n) amount of work, which essentially means every time we match at a new index and every time we backtrack.  It also removes the counting logic that previously translated only 1 out of 1000 CheckTimeout calls into a timeout check; now every CheckTimeout will query the current tick count.

* Address PR feedback

stephentoub committed:
Fix {Last}IndexOfAnyExcept behavior for an empty set of values (#68244)

ViktorHofer committed:
Avoid duplicate Strings.resx import & code-cleanup (#68220)

* Avoid duplicate Strings.resx import & code-cleanup

Fixes https://github.com/dotnet/runtime/issues/68218

Makes assemblies with String.resx resources smaller by avoiding the duplicate resources.

jkotas committed:
Delete legacy HRESULTs and error messages (#68252)

Number of HRESULTs and error messages related to GAC, CAS and other concepts remove in .NET Core
existed just to provide mapping in Marshal.GetExceptionForHR API. They were otherwise unused.

Deleting these HRESULTs is observable behavior change in Marshal.GetExceptionForHR. It should be
acceptable change given that these HRESULTs are never returned by the runtime and very unlikely to
be used by user code.

jkotas committed:
Delete libkrb5-dev from NativeAOT prereqs (#68253)

We load the library dynamically after #55037 so it is not necessary to have the export lib available for the build.

elinor-fung committed:
Add LibraryImportGenerator to ref pack (#68247)

jkotas committed:
Mark OperatingSystem.IsXXX properties as non-versionable (#68250)

It allows them to be inlined by the R2R compiler.

jakobbotsch committed:
Make BSWAP16 nodes normalize upper 16 bits (#67903)

Currently the JIT's constant folding (gtFoldExprConst and VNs
EvalOpSpecialized) assumes that BSWAP16 zero extends into the upper 16
bits. This was not the case, and in fact the behavior of BSWAP16
depended on platform.

Normally this would not be a problem since we always insert normalizing
casts when creating BSWAP16 nodes, however VN was smart enough to remove
this cast in some cases (see the test).

Change the semantics of BSWAP16 nodes to zero extend into the upper 16
bits to match constant folding, and add a small peephole to avoid
inserting this normalization in the common case where it is not
necessary.

Fixes #67723
Subsumes #67726

jakobbotsch committed:
Fix JIT using too wide indirections when returning small structs (#68160)

Fix #58874
Fix #64802
Fix #68157

tannergooding committed:
Exposing ExpM1, Exp2, Exp2M1, Exp10, Exp10M1, LogP1, Log2P1, and Log10P1 (#68251)

* Exposing ExpM1, Exp2, Exp2M1, Exp10, Exp10M1, LogP1, Log2P1, and Log10P1

* Adding tests covering ExpM1, Exp2, Exp2M1, Exp10, Exp10M1, LogP1, Log2P1, and Log10P1

* Ensure Log10P1Test64 is 64-bit only

jakobbotsch committed:
Remove CallArgABIInfo::ArgNum and fill out function headers (#68209)

CallArgABIInfo::ArgNum is essentially unnecessary now that arguments are
stored in order so remove it.

Also address some feedback from the refactoring and fill out some
function headers.

eiriktsarpalis committed:
Add missing RUC annotation to IAsyncEnumerableConverterFactory (#68233)

jkotas committed:
Fix handling of vector instruction sets in NativeAOT compiler (#68249)

The virtual vector instruction used internally by the JIT were leaking out and reported as unsupported

Fixes #68222

vcsjones committed:
Use and support OpenSSL 3 on macOS (#68045)

* Use OpenSSL 3 on macOS

* Update documentation

* Restart CI

akoeplinger committed:
[mono] Preserve FirstChanceExceptionEventArgs ctor (#68235)

The runtime is creating the exception so we need to preserve it.
See https://github.com/xamarin/xamarin-android/issues/6626

While looking at this I noticed that `FirstChanceExceptionEventArgs` is in ExceptionNotification.cs for some reason, fixed the filename to match the type name.

Use DynamicDependency for preserving `FirstChanceExceptionEventArgs` and `UnhandledExceptionEventArgs`

Commit:
Added fixes for tests which are not robust against line endings (#68231)

* Added Fix For System.CodeDom.Tests

* Added Fix for Microsoft.XmlSerializer.Generator tests

* Removed unnecessary File IO

* Substituted LineEndingsHelper.Normalize

Commit:
Fix reading slightly incorrect Zip files (#68106)

* Add test that fails

* fix

* Fix test failing unelevated

* annotate

* overflow

* Apply suggestions from code review

Co-authored-by: Robin Lindner <robin@deeprobin.de>

* Apply suggestions from code review

Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>

Co-authored-by: Robin Lindner <robin@deeprobin.de>
Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>

kunalspathak committed:
fix the srcBufferLength check (#68270)

* fix the srcBufferLength check

* trigger superpmi replay job when changes to superpmi tool

stephentoub committed:
Use generic math to dedup Enumerable.Min/Max (#68183)

* Use generic math to dedup Enumerable.Min/Max

* Address PR feedback

stephentoub committed:
Use generic math to dedup Enumerable.Sum/Average (#68226)

* Use generic math in Enumerable.Sum/Average

* Address PR feedback

elinor-fung committed:
Only set the assembly binder on success when binding to SPCL (#68282)

stephentoub committed:
Fix Docs_Examples_ValidateEmail test (#68294)

The recent change around timeouts appears to have made the timeout used in this test more accurate, which makes it more likely to fire, which is tripping up CI on slower machines.  There's no need for a timeout in this test, though; just delete it.

Commit:
Add retries to ModulesAreDisposedWhenProcessIsDisposed (#68104)

* Add retries to ModulesAreDisposedWhenProcessIsDisposed

* Update src/libraries/System.Diagnostics.Process/tests/ProcessModuleTests.cs

Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>

Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>

stephentoub committed:
Remove dead const from RegexNode (#68297)

joperezr committed:
Use RegexCaseEquivalence table for case-insensitive backreferences (#67977)

strobelt committed:
Add API entry for JsonDocument converter (#60236)

* Add API entry for JsonDocument converter

* Add reference to System.Text.Json

* Add JsonElementConverter from MetadataServices

* Correct converter naming

* Add JsonDocument type to source generation

* include testing

Co-authored-by: Eirik Tsarpalis <eirik.tsarpalis@gmail.com>

AaronRobinsonMSFT committed:
Update TestHandleCollector() test (#68285)

* Update TestHandleCollector() test

Initialize local member in HandleCollector
   since it was the only one not initialized.
Update test to wait for pending finalizers since that is when
   the HandleCollector API will trigger the GC.

Co-authored-by: Elinor Fung <elfung@microsoft.com>

krwq committed:
Add more diagnostics to LookupProperty (more diagnostics to flaky tests) (#68281)

* add more diagnostics to LookupProperty flaky tests

* Fix nullability issue on Release build

akoeplinger committed:
Update Microsoft.DotNet.Helix.Sdk (#68024)

Extracted from https://github.com/dotnet/runtime/pull/67771 to get the fix for https://github.com/dotnet/runtime/issues/67349

Updates the WASM tests to always use net6.0 until we can upgrade to a net7.0 SDK.

Co-authored-by: Premek Vysoky <premek.vysoky@microsoft.com>

jkotas committed:
Add more details to NativeAOT workflow doc (#68308)

Fixes #68257

Commit:
Bump to XHarness w/ android-headless support

Commit:
Force tests to run, for now. Revert when we're happy.

Commit:
Define a Helix queue for Bionic

Commit:
Smoke tests only for now

Commit:
This is fine, we fast path the lookup in managed code on real Android

Commit:
Revert "Smoke tests only for now"

This reverts commit 9d6023e29800eefb49e11fc83c2f8baf151aa6ce.

Commit:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

Commit:
I do not enjoy AzDO YML

Commit:
I still don't enjoy AzDO YML

Commit:
Uh

Commit:
Android exclusions

Commit:
Why do we have two copies of helix-queues-setup?

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Merge branch 'joshield/linux-bionic' of github.com:directhex/runtime into joshield/linux-bionic

directhex committed:
Gotta pass through targetrid to helix.yml somehow?

directhex committed:
Revert "Gotta pass through targetrid to helix.yml somehow?"

This reverts commit 9647c5f46df60681a7801ad6852361c7bac0b726.

directhex committed:
Test a thing from Steve

directhex committed:
Switch to property name used for test runs

directhex committed:
Can we pass through the thing?

directhex committed:
Revert "Can we pass through the thing?"

This reverts commit 9c9d8ddc2153a2cda3752b2ab188bb9214b91d5c.

directhex committed:
SERIOUSLY?!

directhex committed:
Fix up test runner to deal with Android-vs-Bionic HOME issue

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Initial Windows version (for arm64 devices, connected to Windows machines)

directhex committed:
Try again

directhex committed:
Revert some of the last commit

Can't use $OS because we build on Linux

directhex committed:
Try to leverage BrowserHost infra to set HostedOS

directhex committed:
Revert "Try to leverage BrowserHost infra to set HostedOS"

This reverts commit 05a9623f363b19d4725a382a27db581a323ee12a.

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Use IsPosixShell from Helix SDK.

This is expected to still fail, as we are generating the Posix RunTests.sh only

directhex committed:
Whoops, changed during testing, should not have been committed

directhex committed:
Enable overriding of test script shell type generation

directhex committed:
More yaml massage

directhex committed:
Try shuffling things through template hell again

directhex committed:
No need for this twice

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Update xharness (fixes Bionic tests not running on device)

directhex committed:
ICU path on Android 29 varies per architecture, so specify a list

Commit:
Maybe this symlink dir is always usable?

Commit:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

Commit:
Bump to support LinuxBionic exclusions

Commit:
Initial attempt to exclude tests. This should remove failures in System.Text.RegEx

Commit:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

Commit:
Revert "Bump to support LinuxBionic exclusions"

This reverts commit 1da50f643cc84d9572e8d554e3597b44d59df66e.

Commit:
Revert "Revert "Bump to support LinuxBionic exclusions""

This reverts commit b77121029258f332f48e4ee85be9c2db313317d4.

Commit:
runtime compat update

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
This is a computer crime

directhex committed:
Low hanging fruit on getting tests going

directhex committed:
Get going on Linux again?

directhex committed:
Fix typo

directhex committed:
Fixups

directhex committed:
Use SkipOnPlatform to avoid Bionic, not PlatformSpecific

directhex committed:
Add Bionic exclusions for a few more tests

directhex committed:
Conditionally set depsfile, which isn't set in Runtime.Loader.Tests

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Skip a bunch of unpassable tests for Bionic

directhex committed:
More fixes?

directhex committed:
More skips

directhex committed:
More fixes?

directhex committed:
More to skip on Bionic

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
More test fixes

directhex committed:
Maybe fix SSL failures

directhex committed:
More tests

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Revert "This is a computer crime"

This reverts commit 1118077218556bdec2f542668730a59d3a253c6c.

directhex committed:
Test run processor?

directhex committed:
Fix upload dir

directhex committed:
Disambiguate

directhex committed:
Bump xharness to fix failed tests plus all tests on Windows

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
Make WindowsShell usage more specific

directhex commented:
Okay, there's a few small bits & pieces to clean up, but the broad strokes are there (e.g. the on-device build is green and the tests are published), so I'm gonna mark this ready for review. Hopefully we can nail down any lingering issues by the end of the week, and clean up any questionable conditions to everyone's satisfaction.

directhex committed:
Add exclusions for failing tests. Let's get green.

directhex committed:
Fix typo

am11 reviewed (approved):
LGTM. Thank you.

directhex commented on src/native/libs/CMakeLists.txt:
```diff
@@ -162,8 +162,10 @@ if (CLR_CMAKE_TARGET_UNIX OR CLR_CMAKE_TARGET_BROWSER)
+    elseif (FORCE_ANDROID_OPENSSL)
+        add_subdirectory(System.Security.Cryptography.Native)
     else ()
         add_subdirectory(System.Globalization.Native)
```
Seemingly not, the globalization tests seem happy enough

am11 replied:
Ah must be included in a different way https://github.com/dotnet/runtime/blob/f4d03106f43a95d731813d045938955d54345e32/src/mono/CMakeLists.txt#L641 would be nice to converge these to use same mechanism at some point.

directhex replied:
Probably, but out of scope for this PR.

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex committed:
typo

directhex committed:
wat

akoeplinger commented on eng/pipelines/runtime-staging.yml:
```diff
@@ -173,6 +173,47 @@ jobs:
           testRunNamePrefixSuffix: Mono_$(_BuildConfig)
 
+#
+# Build the whole product using Mono for Headless Android (Linux with Bionic libc)
```
nit: let's not call it Headless Android, we don't know if it has a display 😄
```suggestion
# Build the whole product using Mono for Android Linux with Bionic libc
```

akoeplinger commented on eng/testing/BionicRunOnDevice.sh:
```diff
@@ -0,0 +1,69 @@
+if [ $# -eq 0 ]; then
+	echo "Usage: ${currentTest}.sh -r /path/to/dotnet [-l /path/to/preloadlibs]"
+	exit 0
+fi
```
It'd be nice if the argument parsing followed the style in https://github.com/dotnet/runtime/blob/main/eng/testing/RunnerTemplate.sh. Or does the device shell not support while/case/tr etc.?

akoeplinger replied:
there are also a lot of Shellcheck suggestions for this file.

directhex replied:
It's a pretty basic `sh` implementation, I assume BusyBox. There's no Bash at least.

I'd have to test a more sophisticated shell script to see what works.

akoeplinger commented on eng/testing/BionicRunOnDevice.sh:
```diff
@@ -0,0 +1,69 @@
+	export LD_LIBRARY_PATH=$extraLibsDir
+fi
+_RuntimeDir=$(dirname $runtimeExe)
+# Consume OpenSSL as found on Helix
```
let's not mention Helix here, there's nothing tying this to it. if we find openssl then we use it, ideally we'd even pass this in as a parameter to the script?

akoeplinger commented on eng/testing/BionicRunOnDevice.sh:
```diff
@@ -0,0 +1,69 @@
+	esac
+	export LD_LIBRARY_PATH=$_RuntimeDir/openssl/prefab/modules/ssl/libs/android.$_thisArchAndroid:$_RuntimeDir/openssl/prefab/modules/crypto/libs/android.$_thisArchAndroid
+	# Since we're on Helix, we know we want to set the SSL cert dir
+	export SSL_CERT_DIR=/system/etc/security/cacerts
```
I assume we always want this so we can move it out of the if which overrides openssl.

akoeplinger commented on eng/testing/BionicRunnerTemplate.sh:
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env bash
+
```
also a bunch of Shellcheck suggestions for this file.

akoeplinger commented on eng/testing/BionicRunnerTemplate.sh:
```diff
@@ -0,0 +1,61 @@
+    ADDITIONAL_ARGS=${@:5}
+fi
+
+if [[ "x$1" != "x--runtime-path" ]]; then
```
nit: same argument parsing comment as above

akoeplinger commented on eng/testing/tests.targets:
```diff
@@ -1,28 +1,39 @@
     <RunScriptInputName Condition="'$(TargetOS)' == 'Browser' and '$(OS)' != 'Windows_NT' and '$(BuildAOTTestsOnHelix)' == 'true'">WasmRunnerAOTTemplate.sh</RunScriptInputName>
     <RunScriptInputName Condition="'$(TargetOS)' == 'Browser' and '$(OS)' != 'Windows_NT' and '$(BuildAOTTestsOnHelix)' != 'true'">WasmRunnerTemplate.sh</RunScriptInputName>
     <RunScriptInputName Condition="'$(TargetOS)' == 'Browser' and '$(OS)' == 'Windows_NT'">WasmRunnerTemplate.cmd</RunScriptInputName>
+    <RunScriptInputName Condition="$(OutputRid.StartsWith('linux-bionic')) and '$(RunScriptWindowsCmd)' != 'true'">BionicRunnerTemplate.sh</RunScriptInputName>
```
TargetsLinuxBionic

akoeplinger commented on eng/testing/tests.targets:
```diff
@@ -107,6 +118,22 @@
                        OutputPath="$(RunScriptOutputPath)" />
 
     <Exec Condition="'$(TargetOS)' != 'windows' and '$(OS)' != 'Windows_NT'" Command="chmod +x $(RunScriptOutputPath)" />
+    <Copy Condition="$(OutputRid.StartsWith('linux-bionic'))" SourceFiles="$(InnerRunScriptInputPath)" DestinationFiles="$(InnerRunScriptOutputPath)" />
```
TargetsLinuxBionic

akoeplinger commented on eng/testing/tests.targets:
```diff
@@ -107,6 +118,22 @@
+      </_BionicTestEnv>
+    </PropertyGroup>
+
+    <WriteLinesToFile Condition="$(OutputRid.StartsWith('linux-bionic'))" File="$(InnerRunEnvOutputPath)" Overwrite="true" Lines="$(_BionicTestEnv)" Encoding="ascii">
```
TargetsLinuxBionic

akoeplinger commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.props:
```diff
@@ -51,7 +51,7 @@
 
   <!-- Mobile uses a different hosting model, so we don't include the .NET host components. -->
-  <ItemGroup Condition="'$(TargetsMobile)' != 'true'">
+  <ItemGroup Condition="'$(TargetsLinuxBionic)' == 'true' or '$(TargetsMobile)' != 'true'">
```
this seems unnecessary given we don't set TargetsMobile=true for linux bionic

akoeplinger commented on src/libraries/Common/src/Interop/Linux/Interop.Libraries.cs:
```diff
@@ -5,6 +5,9 @@ internal static partial class Interop
     internal static partial class Libraries
     {
+        // Duplicated from Android for Linux Bionic
+        internal const string Liblog = "liblog";
```
liblog.so?

akoeplinger commented on src/libraries/externals.csproj:
```diff
@@ -9,7 +9,7 @@
     <BinPlaceRuntime>false</BinPlaceRuntime>
     <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
-    <UseLiveBuiltDotNetHost Condition="'$(DotNetBuildFromSource)' == 'true' or '$(TargetArchitecture)' == 's390x' or '$(TargetArchitecture)' == 'armv6'">true</UseLiveBuiltDotNetHost>
+    <UseLiveBuiltDotNetHost Condition="'$(DotNetBuildFromSource)' == 'true' or '$(TargetArchitecture)' == 's390x' or '$(TargetArchitecture)' == 'armv6' or $(OutputRid.StartsWith('linux-bionic'))">true</UseLiveBuiltDotNetHost>
```
TargetsLinuxBionic

akoeplinger commented on src/libraries/tests.proj:
```diff
@@ -78,7 +78,7 @@
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetsMobile)' == 'true' or '$(TargetArchitecture)' == 'ARMv6'">
+  <ItemGroup Condition="'$(TargetsMobile)' == 'true' or $(OutputRid.StartsWith('linux-bionic')) or '$(TargetArchitecture)' == 'ARMv6'">
```
TargetsLinuxBionic

akoeplinger commented on src/libraries/tests.proj:
```diff
@@ -129,7 +129,20 @@
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetOS)' == 'Android' and '$(RunDisabledAndroidTests)' != 'true'">
+  <ItemGroup Condition="$(OutputRid.StartsWith('linux-bionic')) and '$(TargetArchitecture)' == 'x64' and '$(RunDisabledAndroidTests)' != 'true'">
```
TargetsLinuxBionic

akoeplinger commented on src/libraries/tests.proj:
```diff
@@ -129,7 +129,20 @@
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.IO.FileSystem\tests\DisabledFileLockingTests\System.IO.FileSystem.DisabledFileLocking.Tests.csproj" />
+  </ItemGroup>
+
+  <ItemGroup Condition="('$(TargetOS)' == 'Android' or $(OutputRid.StartsWith('linux-bionic'))) and '$(RunDisabledAndroidTests)' != 'true'">
```
TargetsLinuxBionic, there are more instances below, I'll stop commenting now

akoeplinger commented on src/native/corehost/corehost.proj:
```diff
@@ -57,7 +57,9 @@
 
-      <BuildArgs>$(Configuration) $(TargetArchitecture) -apphostver "$(AppHostVersion)" -hostver "$(HostVersion)" -fxrver "$(HostResolverVersion)" -policyver "$(HostPolicyVersion)" -commithash "$([MSBuild]::ValueOrDefault('$(SourceRevisionId)', 'N/A'))" -os $(TargetOS)</BuildArgs>
+      <_CoreHostUnixTargetOS>$(TargetOS)</_CoreHostUnixTargetOS>
+      <_CoreHostUnixTargetOS Condition="$(OutputRid.StartsWith('linux-bionic'))">linux-bionic</_CoreHostUnixTargetOS>
```
TargetsLinuxBionic

akoeplinger commented on src/native/corehost/hostmisc/pal.h:
```diff
@@ -83,6 +83,8 @@
 #define FALLBACK_HOST_RID _X("solaris")
 #elif defined(TARGET_LINUX_MUSL)
 #define FALLBACK_HOST_RID _X("linux-musl")
+#elif defined(TARGET_ANDROID)
```
it'd probably be better to introduce a TARGET_LINUX_BIONIC define instead.

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -401,10 +401,13 @@
     </PropertyGroup>
-    <ItemGroup Condition="'$(TargetsAndroid)' == 'true'">
+    <ItemGroup Condition="'$(TargetsLinuxBionic)' == 'true'">
+      <_MonoCPPFLAGS Include="-DANDROID_FORCE_ICU_DATA_DIR" />
```
nit: call this FORCE_ANDROID_ICU_DATA_DIR to make it consistent with FORCE_ANDROID_OPENSSL (or change the other one, I don't have a preference)

akoeplinger replied:
probably better to move this to the src/native CMakeFiles anyway though.

akoeplinger commented on src/native/libs/System.Security.Cryptography.Native/extra_libs.cmake:
```diff
@@ -4,6 +4,25 @@ macro(append_extra_cryptography_libs NativeLibsExtra)
        set(CMAKE_FIND_LIBRARY_SUFFIXES .a)
     endif(CMAKE_STATIC_LIB_LINK)
 
+    # This is evil, bad, and wrong, but good enough to satisfy the build
```
to please policheck:
```suggestion
    # This is bad and wrong, but good enough to satisfy the build
```

akoeplinger commented on src/native/libs/System.Security.Cryptography.Native/opensslshim.c:
```diff
@@ -76,6 +77,14 @@ static void OpenLibraryOnce()
         DlOpen(soName);
     }
 
+#ifdef ANDROID
```
TARGET_ANDROID

akoeplinger commented on src/native/libs/build-native.proj:
```diff
@@ -4,7 +4,9 @@
     <AssemblyName>.NET Runtime</AssemblyName>
-    <_BuildNativeArgs>$(TargetArchitecture) $(Configuration) outconfig $(NetCoreAppCurrent)-$(TargetOS)-$(Configuration)-$(TargetArchitecture) -os $(TargetOS)</_BuildNativeArgs>
+    <_BuildNativeTargetOS>$(TargetOS)</_BuildNativeTargetOS>
+    <_BuildNativeTargetOS Condition="$(OutputRid.StartsWith('linux-bionic'))">linux-bionic</_BuildNativeTargetOS>
```
TargetsLinuxBionic

akoeplinger reviewed (changes requested):
Great jobs, I left a couple comments but this is really close :)

directhex committed:
Merge remote-tracking branch 'origin/main' into joshield/linux-bionic

directhex commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -79,6 +79,10 @@ jobs:
 
+      - ${{ if and(eq(parameters.osGroup, 'Linux'), eq(parameters.osSubGroup, '_bionic')) }}:
+        - name: _osParameter
+          value: -os ${{ parameters.osGroup }} /p:RuntimeOS=linux-bionic /p:OutputRid=linux-bionic-${{ parameters.archType }}
```
Plausibly left over from when we were alternating between Android and Linux as the base OS?

akoeplinger replied:
ok, let's try to make it consistent with musl then

directhex commented on eng/pipelines/common/platform-matrix.yml:
```diff
@@ -185,6 +185,59 @@ jobs:
+        runtimeFlavor: mono
+        # We build on Linux, but the test queue runs Windows, so
+        # we need to override the test script generation
+        runScriptWindowsCmd: true
```
https://github.com/dotnet/runtime/issues/69555

directhex commented on src/installer/pkg/projects/netcoreappRIDs.props:
```diff
@@ -38,6 +38,9 @@
     <OfficialBuildRID Include="linux-musl-arm64">
       <Platform>arm64</Platform>
     </OfficialBuildRID>
+    <OfficialBuildRID Include="linux-bionic-arm64">
```
I haven't the faintest idea, I just put it in because Musl put it in

directhex replied:
I've added a linux-bionic-x64 like for Musl, but not the other two as we aren't building/testing x86/arm for Bionic

am11 replied:
At the bottom, there are `UnofficialBuildRID`s, we should probably add such "I don't care" configurations there.

---

You can build musl-libc, glibc etc. standalone on any platform and link it with your program. But as of today, you _can_ compile bionic-libc on linux-x64 (took me a whole day with almost non-existing practical/working references out in the wild..), but after the compilation, it is unusable and unlinkable (due to missing headers in its INSTALL targets etc.) with anything meaningful. It is all very Android specific.

So bionic-libc is not the same when it comes to flexibility and freedom of use / port.

directhex commented on eng/native/naming.props:
```diff
@@ -6,22 +6,22 @@
   <!-- Add path globs specific to native binaries to exclude unnecessary files from packages. -->
   <Choose>
-    <When Condition="$(PackageRID.StartsWith('win'))">
+    <When Condition="$(OutputRID.StartsWith('win'))">
```
Let's see what happens without.

directhex committed:
Various change suggestions from review by Alex

directhex committed:
additional

directhex committed:
Consistency w/ musl

directhex committed:
Consistency

akoeplinger commented on Directory.Build.props:
```diff
@@ -200,7 +201,8 @@
     <TargetsSolaris Condition="'$(TargetOS)' == 'Solaris'">true</TargetsSolaris>
-    <TargetsLinux Condition="'$(TargetOS)' == 'Linux' or '$(TargetOS)' == 'Android'">true</TargetsLinux>
+    <TargetsLinuxBionic Condition="$(OutputRid.StartsWith('linux-bionic'))">true</TargetsLinuxBionic>
+    <TargetsLinux Condition="'$(TargetOS)' == 'Linux' or '$(TargetOS)' == 'Android' or '$(TargetsLinuxBionic)' == 'true'">true</TargetsLinux>
```
just noticed it: we're using TargetOS=Linux now so TargetsLinuxBionic is redundant here.

directhex committed:
Redundant

directhex committed:
sigh

akoeplinger reviewed (approved):
Thanks, the rest of my comments can be addressed in followup-PRs

directhex committed:
Maybe move a property
